ssl/ssl_lib.c

```diff
@@ -3983,6 +3988,31 @@ static void do_sslkeylogfile(const SSL *ssl, const char *line)
  * via ssl.h.
  */
 
+#ifndef OPENSSL_NO_SSLKEYLOG
+static BIO *get_sslkeylog_bio(const char *keylogfile)
+{
+    int fdno = -1;
+    FILE *fp = NULL;
+
+    fp = fopen(keylogfile, "a");
+    if (fp == NULL)
+        return NULL;
+
+    fdno = fileno(fp);
+    if (fdno == -1) {
+        fclose(fp);
+        return NULL;
+    }
+
+    if (fchmod(fdno, S_IRUSR | S_IWUSR) != 0) {
+        fclose(fp);
+        return NULL;
+    }
```

Comment:
I am sorry but this still has the same TOCTOU race condition I was talking about before.

Why not something like:
```suggestion
    BIO *b;
    
    fdno = open(keylogfile, O_CREAT | O_APPEND,  0600);
    if (fdno < 0)
        return NULL;

    fp = fdopen(fdno, "a");
    if (fp == NULL) {
        close(fdno);
        return NULL;
    }

    if ((b = BIO_new_fp(fp, BIO_CLOSE)) == NULL)
        fclose(fp);
    return b;
```

Reply:
Pushed according slightly tuned your approach. 

The problem I foresee is that parameters of `fdopen` should match the ones we have for `open` and it may make this implementation fragile.

Reply:
Not sure why. `"a"` matches `O_APPEND` and it is in a single function so I do not see any reason they would get out of sync. This is a very much common construct.